src/InstrumentationEngine/BranchTargetInfo.h

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+
+namespace MicrosoftInstrumentationEngine
+{
+    class __declspec(uuid("0E23A44A-6700-4810-889C-8067638C512E"))
+        CBranchTargetInfo : public IUnknown, virtual CModuleRefCount
+    {
+    public:
+        DEFINE_DELEGATED_REFCOUNT_ADDREF(CBranchTargetInfo);
+        DEFINE_DELEGATED_REFCOUNT_RELEASE(CBranchTargetInfo);
+        STDMETHOD(QueryInterface)(_In_ REFIID riid, _Out_ void** ppvObject) override
+        {
+            return ImplQueryInterface(
+                static_cast<IUnknown*>(this),
+                this,
+                riid,
+                ppvObject
+            );
+        }
+    public:
+        static HRESULT GetInstance(_In_ CInstruction* pInstruction, _Outptr_ CBranchTargetInfo** ppResult);
+        static HRESULT GetOrCreateInstance(_In_ CInstruction* pInstruction, _Out_ CBranchTargetInfo** ppResult);
+        static HRESULT SetBranchTarget(_In_ CInstruction* pBranch, _In_ CInstruction* pNewTarget, _In_opt_ CInstruction* pOldTarget);
+        static HRESULT RetargetBranches(_In_ CInstruction* pOriginalInstruction, _In_ CInstruction* pNewInstruction);
+        void Disconnect();
+
+        virtual ~CBranchTargetInfo() = default;
+    private:
+
+        // This is a raw pointer to avoid circular references.
+        CInstruction* m_pInstruction;
+
+        // Using raw pointers for speed. The graph should keep these pointers
+        // alive.
+        std::unordered_set<CInstruction*> m_branches;
```

Comment:
Have you tried some of the extremely large code bases to see what kind of memory impact this has?

Reply:
No, I have not checked the memory impact. It will be linear in the number of branches, though. We could do a scan of mscorlib to find all of the branch instructions, and find an average or something. I don't know the best way to disassemble, though.

Reply:
OK, so I ran an experiment with a program that has about 11000 if instructions, and it looks like the memory is about 10x(pointer_size) per branch instruction. We might be able to make that smaller by reducing the initial size of this set (I can give it a try). We could probably get away with an initial size of 1, since most instructions won't be targeted by many branches.

Reply:
All right. I've updated the constructor to this set, and the map used in CDataContainer to free unused memory correctly, and now the memory impact is zero compared to the baseline of what is in the develop branc.